drgn_tools/meminfo.py

```diff
@@ -0,0 +1,755 @@
+        STRUCT_PAGE_MAX_SHIFT = 6
+        KASAN_SHADOW_SIZE = 0
+        MODULES_VADDR = VA_START + KASAN_SHADOW_SIZE
+        MODULES_VSIZE = SZ_128M
+        MODULES_END = MODULES_VADDR + MODULES_VSIZE
+        VMEMMAP_SIZE = 1 << (VA_BITS - _PAGE_SHIFT - 1 + STRUCT_PAGE_MAX_SHIFT)
+        _VMALLOC_START = MODULES_END
+        _VMALLOC_END = PAGE_OFFSET - PUD_SIZE - VMEMMAP_SIZE - SZ_64K
+
+    else:
+        _IS_X86_64 = True
+
+        # include/linux/numa.h
+        _NODES_SHIFT = 10
+
+        # arch/x86/include/asm/pgtable_64_types.h
+        _PMD_SHIFT = 21
+
+        VMALLOC_SIZE_TB = 32
+        _VMALLOC_START = prog["vmalloc_base"].value_()
+        _VMALLOC_END = _VMALLOC_START + (VMALLOC_SIZE_TB << 40) - 1
+
+        # arch/x86/include/asm/page_types.h
+        _HPAGE_SHIFT = _PMD_SHIFT
+
+    _MAX_NUM_NODES = 1 << _NODES_SHIFT
+    _MAX_NUM_ZONES = prog["__MAX_NR_ZONES"].value_()
+
+    _HPAGE_PMD_SHIFT = _PMD_SHIFT
+    _HPAGE_PMD_ORDER = _HPAGE_PMD_SHIFT - _PAGE_SHIFT
+    _HPAGE_PMD_NR = 1 << _HPAGE_PMD_ORDER
+
+
+def get_active_numa_nodes(prog: Program) -> List[Object]:
+    """
+    Get a list of all active NUMA nodes.
+
+    :returns: a list of ``struct pglist_data *`` objects.
+    """
+    try:
+        active_numa_nodes = prog.cache["active_numa_nodes"]
+    except KeyError:
+        active_numa_nodes = []
+        node_data = prog["node_data"]
+        for i in range(_MAX_NUM_NODES):
+            if node_data[i].value_() != 0x0:
+                active_numa_nodes.append(node_data[i])
+        prog.cache["active_numa_nodes"] = active_numa_nodes
+    return active_numa_nodes
+
+
+def for_each_node_zone(node: Object) -> Iterator[Object]:
+    """
+    Iterate over all memory zones in a NUMA node.
+
+    :returns: Iterator of ``struct zone *`` objects.
+    """
+    node_zones = node.member_("node_zones")
+    for j in range(_MAX_NUM_ZONES):
+        if node_zones[j].value_() != 0x0:
+            yield node_zones[j]
+
+
+def for_each_zone(prog: Program) -> Iterator[Object]:
+    """
+    Iterate over all zones in a system that contains 1+ NUMA nodes.
+
+    :returns: Iterator of ``struct zone *`` objects.
+    """
+    for node in get_active_numa_nodes(prog):
+        for zone in for_each_node_zone(node):
+            yield zone
+
+
+def for_each_block_dev(prog: Program) -> Iterator[Object]:
+    """
+    Iterate over all block devices in a system.
+
+    :returns: Iterator of ``struct block_device *`` objects.
+    """
+    return list_for_each_entry(
+        "struct block_device", prog["all_bdevs"].address_of_(), "bd_list"
+    )
+
+
+def for_each_block_dev_inode(prog: Program) -> Iterator[Object]:
+    """
+    Iterate over all block devices in a system.
+
+    :returns: Iterator of ``struct inode *`` objects.
+    """
+    return list_for_each_entry(
+        "struct inode",
+        prog["blockdev_superblock"].member_("s_inodes").address_of_(),
+        "i_sb_list",
+    )
```

Comment:
My first thought was that these should go into `drgn_tools.block`.

But I now see that they pretty much exist for the purpose of `get_blockdev_pages()`. I may be wrong, but I don't think these have much of a use case beyond that function? I think it would be fine if you were to move these into `get_blockdev_pages()`

Reply:
Yes. These two functions are only used by ```get_blockdev_pages()``` for now.
I don't anticipate any special use case of iterating all block devices' inodes. I am okay to just keep them inline because they are pretty short helper functions. If we do need them later, we can move this to ```drgn_tools.block```.